tests/test_geometry.py

```diff
@@ -51,3 +55,240 @@ def test_check_if_geometry_is_valid():
+    region.add_entity(geometry.Line((0, 0), (1, 1)))
+
+    entity = geometry.Line((-2, 2), (2, 3))
+    region.insert_entity(0, entity)
+
+    assert (len(region.entities) == 2) & (region.entities[0] == entity)
+
+
+def test_region_insert_polyline():
+    region = geometry.Region()
+    region.insert_polyline(
+        0,
+        [
+            geometry.Line((0, 0), (1, 1)),
+            geometry.Line((1, 1), (1, 0)),
+            geometry.Line((1, 0), (0, 0)),
+        ],
+    )
+
+    assert len(region.entities) == 3
+
+
+def test_region_remove_entity():
+    region = geometry.Region()
+    region.insert_polyline(
+        0,
+        [
+            geometry.Line((0, 0), (1, 1)),
+            geometry.Line((1, 1), (1, 0)),
+            geometry.Line((1, 0), (0, 0)),
+        ],
+    )
+    region.remove_entity(region.entities[1])
+
+    assert len(region.entities) == 2
+
+
+def test_region_from_json():
+    raw_region = {
+        "name": "test_region",
+        "material": "copper",
+        "colour": {"r": 240, "g": 0, "b": 0},
+        "area": 5.1,
+        "centroid": {"x": 0.0, "y": 1.0},
+        "region_coordinate": {"x": 0.0, "y": 1.1},
+        "duplications": 10,
+        "entities": [],
+    }
+
+    test_region = geometry.Region()
+    test_region.name = "test_region"
+    test_region.material = "copper"
+    test_region.colour = (240, 0, 0)
+    test_region.area = 5.1
+    test_region.centroid = (0.0, 1.0)
+    test_region.region_coordinate = (0.0, 1.1)
+    test_region.duplications = 10
+    test_region.entities = []
+
+    region = geometry.Region()
+    region._from_json(raw_region)
+
+    assert region.name == test_region.name
+    assert region.material == test_region.material
+    assert region.colour == test_region.colour
+    assert region.area == test_region.area
+    assert region.centroid == test_region.centroid
+    assert region.region_coordinate == test_region.region_coordinate
+    assert region.duplications == test_region.duplications
+    assert region.entities == test_region.entities
+
+
+def test_region_to_json():
+    raw_region = {
+        "name": "test_region",
+        "material": "copper",
+        "colour": {"r": 240, "g": 0, "b": 0},
+        "area": 5.1,
+        "centroid": {"x": 0.0, "y": 1.0},
+        "region_coordinate": {"x": 0.0, "y": 1.1},
+        "duplications": 10,
+        "entities": [],
+    }
+
+    test_region = geometry.Region()
+    test_region.name = "test_region"
+    test_region.material = "copper"
+    test_region.colour = (240, 0, 0)
+    test_region.area = 5.1
+    test_region.centroid = (0.0, 1.0)
+    test_region.region_coordinate = (0.0, 1.1)
+    test_region.duplications = 10
+    test_region.entities = []
+
+    assert test_region._to_json() == raw_region
+
+
+def test_line_get_coordinate_from_percentage_distance():
+    line = geometry.Line((0, 0), (1, 1))
+
+    x, y = line.get_coordinate_from_percentage_distance(0, 0, 0.5)
+    assert (x, y) == (0.5, 0.5)
```

Comment:
maybe worth having a test case where x/y are different. Otherwise could miss typo where we return in wrong order